packages/utils/crates/utils-types/src/sma.rs

```diff
@@ -1,56 +1,67 @@
 use ringbuffer::{AllocRingBuffer, RingBuffer};
 use std::fmt::{Display, Formatter};
+use std::marker::PhantomData;
 use std::ops::{Add, Div};
 
 /// Simple Moving Average trait.
 /// The second-most useful filter type, bested only by coffee filters.
-pub trait SMA<T>
-where
-    T: for<'a> Add<&'a T, Output = T> + Div<T, Output = T> + Clone + From<u32> + Display + Default,
-{
+pub trait SMA<T> {
     /// Adds a sample.
     fn add_sample(&mut self, sample: T);
 
     /// Calculates the moving average value.
-    fn get_average(&self) -> T;
+    /// Returns `None` if no samples were added.
+    fn get_average(&self) -> Option<T>;
 
     /// Returns the window size.
     fn window_size(&self) -> usize;
 
     /// Returns the number of elements in the window.
     /// This value is always between 0 and `window_size()`.
     fn len(&self) -> usize;
+
+    /// Indicates whether the window is fully occupied
+    /// with samples.
+    fn is_window_full(&self) -> bool {
+        self.len() == self.window_size()
+    }
+
+    /// Indicates whether there are no samples.
+    fn is_empty(&self) -> bool;
 }
 
 /// Basic implementation of Simple Moving Average (SMA).
 /// The maximum window size is bound by 2^32 - 1.
 /// Useful mainly for floating-point types, as it does not accumulate floating point error with each sample,
 /// but requires `O(N)` of memory and `O(N)` for average computation, `N` being window size.
+/// The divisor argument `D` is used only for such types `T` that do not implement `From<u32>` (such as `Duration`,...).
 #[derive(Clone, Debug, PartialEq)]
-pub struct NoSumSMA<T>
+pub struct NoSumSMA<T, D = T>
 where
-    T: for<'a> Add<&'a T, Output = T> + Div<T, Output = T> + Clone + From<u32> + Display + Default,
+    T: Clone,
 {
     window: AllocRingBuffer<T>,
+    _div: PhantomData<D>,
 }
 
-impl<T> SMA<T> for NoSumSMA<T>
+impl<T, D> SMA<T> for NoSumSMA<T, D>
 where
-    T: for<'a> Add<&'a T, Output = T> + Div<T, Output = T> + Clone + From<u32> + Display + Default,
+    T: Add<T, Output = T> + Div<D, Output = T> + Clone,
+    D: From<u32>,
 {
     fn add_sample(&mut self, sample: T) {
         self.window.push(sample);
     }
 
-    fn get_average(&self) -> T {
+    fn get_average(&self) -> Option<T> {
         if !self.window.is_empty() {
-            let mut ret = T::default();
-            for v in self.window.iter() {
+            let mut ret = self.window[0].clone();
+            for v in self.window.iter().skip(1).cloned() {
                 ret = ret + v;
             }
-            ret / (self.window.len() as u32).into()
+            Some(ret / D::from(self.window.len() as u32))
         } else {
-            Default::default()
+            None
         }
```

Comment:
This is an accumulating fold, a.k.a `reduce`:
```suggestion
        self.window.iter().reduce(|acc, value| acc + value).map(|reduced| reduced / self.window().len())
```

Reply:
All the SMA related improvements were resolved in: https://github.com/hoprnet/hoprnet/pull/5744